src/Query/QueryBuilder.php

```diff
@@ -1259,6 +1267,34 @@ public function resetOrderBy(): self
         return $this;
     }
 
+    /**
+     * Adds a CTE to a select query.
+     *
+     * @param string $name         The name of the CTE.
+     * @param self   $queryBuilder The query builder to use within the CTE.
+     *
+     * @return $this This QueryBuilder instance.
+     */
+    public function with(string $name, self $queryBuilder): self
+    {
+        if (! $this->connection->getDatabasePlatform()->supportsCTEs()) {
+            throw new QueryException('CTEs are not supported by this database platform.');
+        }
+
+        if ($queryBuilder->hasCTEs()) {
+            throw new QueryException('CTEs cannot be nested.');
+        }
+
+        $this->with[$name] = $queryBuilder;
+
+        return $this;
+    }
+
+    private function hasCTEs(): bool
+    {
+        return 0 < count($this->with);
```

Comment:
```suggestion
        return $this->with !== [];
```

Reply:
I agree that the Yoda-style condition looks odd but the `!== []` looks like a PHP-ism.

Generally, in order to compare two arrays, one first need to compare their lengths and, if they are equal, iterate both arrays until a non-equal element is found.

In this case, checking if `count($array) === 0` looks like a more meaningful alternative because we don't really need to the full-fledged/generic comparison algorithm.